File: Adhiti_Shaili/Day2/Maximum Product Subarray.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution{
public:
	// Function to find maximum product subarray
	long long maxProduct(int *arr, int n) {
	    long long pprod=1,ans=1,nprod=1;
	    for(int i=0;i<n;i++)
	    {
	        if(arr[i]<0)
	        {
	            swap(pprod,nprod);
	        }
	        
	        pprod=max((long long)arr[i],(long long)arr[i]*pprod);
	        nprod=min((long long)arr[i],(long long)arr[i]*nprod);
	        ans=max(ans,pprod);
	        
	    }
	    return ans;
	}
};

int main() {
    int t;
    cin >> t;
    while (t--) {
        int n, i;
        cin >> n;
        int arr[n];
        for (i = 0; i < n; i++) {
            cin >> arr[i];
        }
        Solution ob;
        auto ans = ob.maxProduct(arr, n);
        cout << ans << "\n";
    }
    return 0;
} 

File: Ajaz_Ahmed/Day2/Maximum product subarray.py
#User function Template for python3
class Solution:

	# Function to find maximum
	# product subarray
	def maxProduct(self,arr, n):
        max_ending = min_ending = arr[0]
        max_so_far = arr[0]
        for i in range(1,len(arr)):
            temp = max_ending
            max_ending = max(arr[i], max(arr[i]*temp, arr[i]*min_ending))
            min_ending = min(arr[i], min(arr[i]*temp, arr[i]*min_ending))
            max_so_far = max(max_ending, max_so_far)
        return max_so_far     

#{ 
#  Driver Code Starts
#Initial Template for Python 3



if __name__ == '__main__':
    tc = int(input())
    while tc > 0:
        n = int(input())
        arr = list(map(int, input().strip().split()))
        ob = Solution()
        ans = ob.maxProduct(arr, n)
        print(ans)
        tc -= 1

# } Driver Code Ends
